### PR TITLE
Javalab: Create BackpackClientApi

### PR DESCRIPTION
**Pull request by molly-moen**
Create a client api we can use in javalab to save student code to the backpack. This PR adds the client api to the Javalab redux store for now and will be used once the commit modal is ready for it.

Since we only want students to get a backpack channel token on their first backpack save, we send over their existing backpack channel token if it exists as part of the lab configuration. If they don't yet have a channel id when we save for the first time, we fetch it via the `/backpacks/channel` api call. When saving student code, we retry failures once per file, then surface a failure to the student.

## Links

- spec: [Backpack dev doc](https://docs.google.com/document/d/1Y01X2_rrVt8f4y8yHpm3BBpq31Ra9CSkOLYycwQ5or8/edit#heading=h.of78enh7nbo1)
- jira ticket: [CSA-523](https://codedotorg.atlassian.net/browse/CSA-523)

## Testing story
Added unit tests for the new client api and tested locally.

## Follow-up work
Connect the client api to the commit and save modal.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/components/backpack/BackpackClientApi.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        filename,
+        fileContents,
+        // onError: retry first failure, then call onError method on second failure
+        _ =>
+          this.writeSingleFileToBackpack(
+            filename,
+            fileContents,
+            onError,
+            onSuccess
+          ),
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/components/backpack/BackpackClientApi.js]: Hmm, would it be better to pass a retry count and roll the retry logic into `writeSingleFileToBackpack`? This works for a single retry but would we need to configure multiple retries down the road? Fine to keep it this way though if you think we only need to support a single retry for now.

File: apps/src/code-studio/components/backpack/BackpackClientApi.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      this.fetchChannelId(() =>
+        this.saveFilesHelper(filesJson, filenames, onError, onSuccess)
+      );
+    } else {
+      this.saveFilesHelper(filesJson, filenames, onError, onSuccess);
+    }
+  }
+
+  saveFilesHelper(filesJson, filenames, onError, onSuccess) {
+    this.filesToUpload = filenames;
```
Comment (sanchitmalhotra126) [inline apps/src/code-studio/components/backpack/BackpackClientApi.js:45]: Do we need to guard against back-to-back calls to `saveFilesHelper`? We could potentially overwrite `this.filesToUpload` while file writing is in progress right?

**Review by sanchitmalhotra126 — COMMENTED** [review]
Couple questions but nothing major, otherwise LGTM

File: apps/src/code-studio/components/backpack/BackpackClientApi.js (hunk 2 above)
Comment (molly-moen) [reply apps/src/code-studio/components/backpack/BackpackClientApi.js:45]: yep, will update

File: apps/src/code-studio/components/backpack/BackpackClientApi.js (hunk 1 above)
Comment (molly-moen) [reply apps/src/code-studio/components/backpack/BackpackClientApi.js]: I think that's a good idea and it will make things a bit cleaner.

**Review by sanchitmalhotra126 — APPROVED** [review]

**Pull request merged**